`gardener-node-agent`: Persist applied changes after each step

maboehm opened the pull request:
**How to categorize this PR?**
/area robustness
/kind enhancement

**What this PR does / why we need it**:
Persists the `operationSystemConfigChanges` to disk and updates it after each step (for example after applying a file). If the OSC is changed, the changes are always re-computed.

So the possible scenarios are:
- GNA runs for the first time: no `oscc.yaml`, no `last-applied-osc.yaml`: Apply entire OSC
- GNA encounters an error: `oscc.yaml` exists on disk, the checksum still matches: resume with the changes in `oscc.yaml`
- GNA is stuck, because the OSC (Version 2) is broken, then the OSC is updated (Version 3): `oscc.yaml` exists, but its checksum does not match. Compute changes as before.

**Which issue(s) this PR fixes**:
This is part of the [Hackathon December 2024](https://github.com/gardener-community/hackathon/tree/main/2024-12_Schelklingen)

**Special notes for your reviewer**:

**Release note**:
```feature operator
`gardener-node-agent` now persists its applied changes after each step when reconciling the OSC. This should avoid unnecessary work and systemd unit restarts.
```

maboehm committed:
Remember work-in-progress in node agent

maboehm committed:
add integration tests

oliver-goetz commented on pkg/nodeagent/controller/operatingsystemconfig/osc_changes.go:
```diff
@@ -0,0 +1,211 @@
+// SPDX-FileCopyrightText: SAP SE or an SAP affiliate company and Gardener contributors
```
```suggestion
// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
```

oliver-goetz commented on pkg/nodeagent/controller/operatingsystemconfig/osc_changes.go:
```diff
@@ -0,0 +1,211 @@
+	Units           units      `json:"units"`
+	Files           files      `json:"files"`
+	Containerd      containerd `json:"containerd"`
+	RestartRequired bool       `json:"restartSelf"`
```
```suggestion
	RestartRequired bool       `json:"restartRequired"`
```

oliver-goetz commented on pkg/nodeagent/controller/operatingsystemconfig/osc_changes.go:
```diff
@@ -0,0 +1,211 @@
+	if err != nil {
+		return err
+	}
+	return o.fs.WriteFile(operatingSystemConfigChangesFilePath, out, 0644)
```
```suggestion
	return o.fs.WriteFile(operatingSystemConfigChangesFilePath, out, 0600)
```

oliver-goetz commented on pkg/nodeagent/controller/operatingsystemconfig/osc_changes.go:
```diff
@@ -0,0 +1,211 @@
+	})
+
+	o.lock.Unlock()
+	return o.persist()
```
If another caller is waiting for the lock, this could persist something different than the content computed in this function.
Same for the other functions.

maboehm replied:
yea, that is true, but `persist()` is also called directly (once), which is why I also added the lock there (and need to release it here). But I guess you are right, it's probably better that `persist()` does not wait for the lock at all and assumes the caller (usually a `complete*()` function is holding the lock.

oliver-goetz commented on pkg/nodeagent/controller/operatingsystemconfig/reconciler.go:
```diff
@@ -41,7 +42,10 @@ import (
-const lastAppliedOperatingSystemConfigFilePath = nodeagentv1alpha1.BaseDir + "/last-applied-osc.yaml"
+const (
+	lastAppliedOperatingSystemConfigFilePath = nodeagentv1alpha1.BaseDir + "/last-applied-osc.yaml"
+	operatingSystemConfigChangesFilePath     = nodeagentv1alpha1.BaseDir + "/oscc.yaml"
```
```suggestion
	operatingSystemConfigChangesFilePath     = nodeagentv1alpha1.BaseDir + "/osc-changes.yaml"
```

oliver-goetz reviewed (commented):
Nice PR, thanks 🚀

maboehm committed:
PR feedback, adjust lock handling

maboehm committed:
Create deepcopy to avoid changes

in https://github.com/gardener/gardener/pull/11015 the code was changed to update the `OSC` and persist that to disk, instead of just the raw bytes read from the secret. This caused conflicts with the operatingSystemConfigChanges because some deletions affected both.

rfranzke commented on pkg/nodeagent/controller/operatingsystemconfig/changes.go:
```diff
@@ -45,46 +47,38 @@ func extractOSCFromSecret(secret *corev1.Secret) (*extensionsv1alpha1.OperatingS
-	registries containerdRegistries
-}
+	if oldChanges.OSCChecksum == changes.OSCChecksum {
+		log.Info("Found OSC changes on disk, remaining work",
```
```suggestion
		log.Info("Found previously computed OperatingSystemConfig changes on disk, remaining work",
```

rfranzke commented on pkg/nodeagent/controller/operatingsystemconfig/changes.go:
```diff
@@ -45,46 +47,38 @@ func extractOSCFromSecret(secret *corev1.Secret) (*extensionsv1alpha1.OperatingS
-	desired []extensionsv1alpha1.RegistryConfig
-	deleted []extensionsv1alpha1.RegistryConfig
-}
+	log.Info("OSC changes checksum did not match, computing new changes")
```
```suggestion
	log.Info("OperatingSystemConfig changes checksum does not match, computing new changes")
```

rfranzke commented on pkg/nodeagent/controller/operatingsystemconfig/changes.go:
```diff
@@ -97,25 +91,33 @@ func computeOperatingSystemConfigChanges(fs afero.Afero, newOSC *extensionsv1alp
 		}
 
 		var unitChanges []changedUnit
+		var unitCommands []unitCommand
```
Nit: Use `var ()` block

rfranzke commented on pkg/nodeagent/controller/operatingsystemconfig/containerd.go:
```diff
@@ -434,15 +442,22 @@ func addRegistryToContainerdFunc(ctx context.Context, log logr.Logger, registryC
+	if err := tplContainerdHosts.Execute(f, values); err != nil {
+		return err
+	}
+	log.Info("Configured registryConfig", "upstream", registryConfig.Upstream)
```
```suggestion
	log.Info("Configured registry config", "upstream", registryConfig.Upstream)
```

rfranzke commented on pkg/nodeagent/controller/operatingsystemconfig/osc_changes.go:
```diff
@@ -0,0 +1,214 @@
+	lock sync.Mutex
+	fs   afero.Afero
+
+	OSCChecksum     string     `json:"oscChecksum"`
```
```suggestion
	OperatingSystemConfigChecksum     string     `json:"operatingSystemConfigChecksum"`
```

rfranzke commented on pkg/nodeagent/controller/operatingsystemconfig/osc_changes.go:
```diff
@@ -0,0 +1,214 @@
+	Units           units      `json:"units"`
+	Files           files      `json:"files"`
+	Containerd      containerd `json:"containerd"`
+	RestartRequired bool       `json:"restartRequired"`
```
```suggestion
	MustRestartNodeAgent bool       `json:"mustRestartNodeAgent"`
```

rfranzke commented on pkg/nodeagent/controller/operatingsystemconfig/osc_changes.go:
```diff
@@ -0,0 +1,214 @@
+
+type containerd struct {
+	// ConfigFileChange tracks if the config file of containerd will change, so that GNA can restart the unit.
+	ConfigFileChange bool `json:"configFileChange"`
```
```suggestion
	ConfigFileChanged bool `json:"configFileChanged"`
```

rfranzke commented on pkg/nodeagent/controller/operatingsystemconfig/osc_changes.go:
```diff
@@ -0,0 +1,214 @@
+func (o *operatingSystemConfigChanges) persist() error {
+	out, err := yaml.Marshal(o)
+	if err != nil {
+		return err
```
```suggestion
		return fmt.Errorf("failed marshalling the changes into YAML: %w", err)
```

rfranzke commented on pkg/nodeagent/controller/operatingsystemconfig/osc_changes.go:
```diff
@@ -0,0 +1,214 @@
+	return o.fs.WriteFile(operatingSystemConfigChangesFilePath, out, 0600)
+}
+
+func (o *operatingSystemConfigChanges) setRestartRequired(restart bool) error {
```
```suggestion
func (o *operatingSystemConfigChanges) setMustRestartNodeAgent(restart bool) error {
```

rfranzke commented on pkg/nodeagent/controller/operatingsystemconfig/osc_changes.go:
```diff
@@ -0,0 +1,214 @@
+func loadOSCChanges(fs afero.Afero) (*operatingSystemConfigChanges, error) {
+	raw, err := fs.ReadFile(operatingSystemConfigChangesFilePath)
+	if err != nil {
+		return nil, err
```
```suggestion
		return nil, fmt.Errorf("failed reading changes file %s: %w", operatingSystemConfigChangesFilePath, err)
```

rfranzke commented on pkg/nodeagent/controller/operatingsystemconfig/osc_changes.go:
```diff
@@ -0,0 +1,214 @@
+	changes := operatingSystemConfigChanges{}
+	if err := yaml.Unmarshal(raw, &changes); err != nil {
+		return nil, err
+	}
```
```suggestion

	changes := operatingSystemConfigChanges{}
	if err := yaml.Unmarshal(raw, &changes); err != nil {
		return nil, fmt.Errorf("failed unmarshalling the changes: %w", err)
	}

```

rfranzke commented on pkg/nodeagent/controller/operatingsystemconfig/reconciler.go:
```diff
@@ -45,7 +46,10 @@ import (
-const lastAppliedOperatingSystemConfigFilePath = nodeagentv1alpha1.BaseDir + "/last-applied-osc.yaml"
+const (
+	lastAppliedOperatingSystemConfigFilePath = nodeagentv1alpha1.BaseDir + "/last-applied-osc.yaml"
+	operatingSystemConfigChangesFilePath     = nodeagentv1alpha1.BaseDir + "/osc-changes.yaml"
```
```suggestion
	lastComputedOperatingSystemConfigChangesFilePath     = nodeagentv1alpha1.BaseDir + "/last-computed-osc-changes.yaml"
```

WDYT?

rfranzke reviewed (commented):
Thanks @maboehm, I have only a few small suggestions :)

maboehm committed:
Address naming, error improvements from PR review

rfranzke reviewed (commented):
/lgtm
/approve
/retest
